central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, val := range indicators {
+		key := fmt.Sprintf("%s %s %s %s %s",
+			val.GetContainerName(),
+			val.GetPodId(),
+			val.GetSignal().GetName(),
+			val.GetSignal().GetArgs(),
+			val.GetSignal().GetExecFilePath(),
+		)
+
+		// A bit of paranoia is always good
+		if old, ok := indicatorsMap[key]; ok {
+			log.Warnf("An indicator %s is already present, overwrite with %s",
+				old.GetId(), val.GetId())
+		}
+
+		indicatorsMap[key] = val
+		indicatorIds = append(indicatorIds, val.GetId())
+	}
+
+	return indicatorsMap, indicatorIds, nil
+}
+
+// OpenClosedPLOPs is a convenient type alias to use in PLOP normalization
+type OpenClosedPLOPs struct {
+	open   []*storage.ProcessListeningOnPortFromSensor
+	closed []*storage.ProcessListeningOnPortFromSensor
+}
+
+// normalizePLOPs
+//
+// In the batch of PLOP events there could be many open & close events for the
+// same combination of port, protocol, process. Find and squash them into a
+// single event.
+//
+// Open/close state will be calculated from the totall number of open/close
+// events in the batch, assuming every single open will eventually be followed
+// by close. In this way in the case of:
+// * Out-of-order events, we would be able to establish correct status
+// * Pairs split across two batches, the status will be correct after processing both batches
+// * Lost events, the status will be incorrect
+// Another alternative would be to set the status based on the final PLOP
+// event, which will produce the same results for case 2 and 3. But such
+// approach will produce incorrect status in the case 1 as well, so counting
+// seems to be more preferrable.
+func normalizePLOPs(
+	plops []*storage.ProcessListeningOnPortFromSensor,
+) []*storage.ProcessListeningOnPortFromSensor {
+
+	normalizedMap := map[string]OpenClosedPLOPs{}
+	normalizedResult := []*storage.ProcessListeningOnPortFromSensor{}
+
+	for _, val := range plops {
+		key := fmt.Sprintf("%d %d %s",
+			val.GetProtocol(),
+			val.GetPort(),
+			getProcesUniqueKey(val),
+		)
+
+		if prev, ok := normalizedMap[key]; ok {
+
+			if val.GetCloseTimestamp() == nil {
+				prev.open = append(prev.open, val)
+			} else {
+				prev.closed = append(prev.closed, val)
+			}
+
+			normalizedMap[key] = prev
+
+		} else {
+
+			newValue := OpenClosedPLOPs{
+				open:   []*storage.ProcessListeningOnPortFromSensor{},
+				closed: []*storage.ProcessListeningOnPortFromSensor{},
+			}
+
+			if val.GetCloseTimestamp() == nil {
+				newValue.open = append(newValue.open, val)
+			} else {
+				newValue.closed = append(newValue.closed, val)
+			}
+
+			normalizedMap[key] = newValue
+		}
+	}
+
+	for _, value := range normalizedMap {
+		nOpen := len(value.open)
+		nClosed := len(value.closed)
+
+		// Take the last open PLOP if there are more open in total, otherwise
+		// the last closed PLOP
+		if nOpen > nClosed {
```

Comment:
I might be missing something.  How does this handle the case of PLOPs that are out of order?  This seems to return the last one you processed which would assume that `plops` was passed in ordered.  May not matter so much for open, but could pose issues with closed and the closedTimestamp.

Reply:
See above response to similar comment.